### PR TITLE
torchdata datapipes drop addition

### LINKED ISSUE #656: Ability to manipulate columns and fields
### 🚀 The feature

- [x] Add ability to drop specific column / field

For `list` and `tuple`
```python
list(dp) # [ (0, 1, 2), (3, 4, 5), (6,7, 8) ]
dp = dp.drop(1)
list(dp) # [ (0, 2), (3, 5), (6, 8) ]
```

Similar for `dict` types
```python
list(dp) # [ {a:1, b:2}, {a:3, b:4}, {a:5, b:6}]
dp = dp.drop('a')
list(dp) # [ {b:2}, {b:4}, {b:6}]
```

- [ ] Add ability to slice fields

For `list` and `tuple`
```python
list(dp) # [ (0, 1, 2, 0, 1, 2), (3, 4, 5, 3,4,5), (6,7, 8,6,7,8) ]
dp = dp[1:-1]
list(dp) # [ ( 1, 2, 0, 1), ( 4, 5, 3,4), (7, 8,6,7) ]
```

For `dict`
```python
list(dp) # [ {a:1, b:2, c:3, d:4}, {a:3, b:4, c:3, d:4}, {a:5, b:6, c:3, d:4}]
dp = dp['a','b']
list(dp) #  [ {a:1, b:2}, {a:3, b:4}, {a:5, b:6}]
```

- [ ] Add ability to flatten structures

```python
list(dp) # [ (1, (2,3), 4), (5, (6,7), 8) ]
dp = dp.flatten(1)
list(dp) # [ (1, 2, 3,4), (5,6,7,8) ]
```

Note: No arguments `flatten()` should cover use-case from #648
Note: Exception if structures are different length 

```python
list(dp) # [ { a: 1, b: {e: 2, f:3 }, c: 4} , { a: 1, b: {e: 2, f:3 }, c: 4} ]
dp = dp.flatten('b')
list(dp) # [ { a: 1, e: 2, f:3 , c: 4}  , { a: 1, e: 2, f:3, c: 4}]
```
Note: Exception if keys overlaps

CC @NivekT @ejguan

### PR DESCRIPTION
**Pull request by dbish**
Please read through our [contribution guide](https://github.com/pytorch/data/blob/main/CONTRIBUTING.md) prior to
creating your pull request.

- Note that there is a section on requirements related to adding a new DataPipe.

Fixes #656 

### Changes
part 1 of the feature requests to manipulate datapipe columns here: https://github.com/pytorch/data/issues/656
this adds a drop functionality to iter datapipes

### CODE REVIEW

**Review by ejguan — APPROVED** [review]
Overall, LGTM. Thank you for adding this DataPipe. Could you please fix the lint Error?
For formatting, you can follow the instruction here: https://github.com/pytorch/data/blob/main/CONTRIBUTING.md#code-style. And, run `pre-commit run --all-files` to execute auto formatting.
For mypy, you can execute `mypy --config-file mypy.ini` to see the Error.

**Review by NivekT — APPROVED** [review]
LGTM! Just one small step to add this DataPipe to the documentation.

Add `Dropper` into `docs/source/torchdata.datapipes.iter.rst`. I think the category "Selecting" is the most fitting, you can put it right before "Filter".

File: torchdata/datapipes/iter/transform/callable.py [hunk 1]
```diff
@@ -137,3 +138,60 @@ def __iter__(self) -> Iterator[T_co]:
...
+            self.indices = {indices}
+
+    def __iter__(self) -> Iterator[T_co]:
+        for old_item in self.datapipe:
+            if isinstance(old_item, tuple):
+                new_item = tuple(x for i, x in enumerate(old_item) if i not in self.indices)
+            elif isinstance(old_item, list):
+                new_item = [x for i, x in enumerate(old_item) if i not in self.indices]
+            elif isinstance(old_item, dict):
+                new_item = {k: v for (k, v) in old_item.items() if k not in self.indices}
```
[the remark this answers is not in the document]
Comment (dbish) [reply torchdata/datapipes/iter/transform/callable.py:182]: We can raise a warning by doing something like this and running a quick check over the indices:
`#filter indices all present check
try:
    for i in self.indices:
        old_item[i]
except (IndexError, KeyError):
    warnings.warn(
        "At least one index in the filter is not present in the item being returned,"
        " please be aware that expected columns/keys may be missing."
    )`

[the remark this answers is not in the document]
Comment (ejguan) [reply torchdata/datapipes/iter/transform/callable.py:182]: SGTM.

**Review by ejguan — APPROVED** [review]

**Pull request closed without merging**